DOCS-223 CLC Overview (#95)
hazelcast/hazelcast-commandline-client: docs/modules/ROOT/pages/overview.adoc

= Hazelcast Command-Line Client (CLC)
:url-github-clc: https://github.com/hazelcast/hazelcast-cloud-cli/blob/master/README.md 
:description: You can use the Hazelcast Command Line Client (CLC) to connect to and interact with clusters on Hazelcast Platform and {hazelcast-cloud}, direct from the command line or through scripts.

{description}

The Hazelcast CLC simplifies the client setup process for Unix-like local environments. Within minutes of installation, you can start to perform common tasks on clusters.

== Use Cases

Example use cases for the Hazelcast CLC.

=== Basic Cluster Management for Hazelcast Platform

Perform xref:hzc-cluster.adoc[administration tasks] on multiple clusters without writing code. Instead, use <<interactive-mode, interactive mode>> to explore available commands.

Check or change the state of a cluster, update a configuration option, then switch to another cluster, all from the command line. For maximum flexibility, you can set cluster connection details at the command line or using a xref:configuration.adoc[custom configuration file]. 

> NOTE: {hazelcast-cloud} has a separate link:{url-github-clc}[CLI for cluster management].

Reviewer: I can not see the image on preview deployment. Maybe it is expected but just wanted to point out
https://deploy-preview-95--eclectic-sawine-19fcf1.netlify.app/docs/clc/0.1/overview.html

=== Non-Interactive Mode

This is the default mode for the Hazelcast CLC. You need to manually enter and execute commands at the command line. This mode is useful for scripting.

[source,bash,subs="attributes+"]
----
hzc map --cluster-name cluster1 --name myMap put --key myKey --value myValue
----

[[interactive-mode]]
=== Interactive Mode

Interactive mode uses auto-completion to display matching commands, subcommands, and parameters. This mode is useful for exploring CLC commands, and for manual tasks. 

image:ROOT:hzc-interactive-screenshot.png[HZC interactive]

In interactive mode, you can run the xref:hzc-map.adoc[`hzc map use` command] to avoid re-typing the object name on each command. In this example, the developer only needed to type `m1` once.

----
hzc
Connecting to the cluster ...
hzc localhost:5701@dev> map use m1
hzc localhost:5701@dev&m:m1> map put -k k1 -v v1
hzc localhost:5701@dev&m:m1> map get -k k1
v1
hzc localhost:5701@dev&m:m1>
----

You can xref:install-clc.adoc[enable interactive mode] for bash and zsh shells after installation. To start Hazelcast CLC in interactive mode, do the following:

[source,bash,subs="attributes+"]
----
hzc
----

> To return to non-interactive mode, use either of the following methods:

Reviewer: ```suggestion
To exit, use either of the following methods:
```
I think this one is correct

= Hazelcast Command-Line Client (CLC)
:url-github-clc: https://github.com/hazelcast/hazelcast-cloud-cli/blob/master/README.md 

> :description: You can use Hazelcast Command Line Client (CLC) to connect to and interact with clusters on Hazelcast Platform and Hazelcast Viridian, direct from the command line or through scripts.

Author: Updated as requested.

> :description: You can use Hazelcast Command Line Client (CLC) to connect to and interact with clusters on Hazelcast Platform and Hazelcast Viridian, direct from the command line or through scripts.

Reviewer 2: ```suggestion
:description: You can use the Hazelcast Command Line Client (CLC) to connect to and interact with clusters on Hazelcast Platform and Hazelcast Viridian, direct from the command line or through scripts.
```

Author: Updated as requested.

{description}

> Hazelcast CLC simplifies the client setup process for Unix-like local environments. Within minutes of installation, you can start to perform common tasks on clusters.

Reviewer 2: ```suggestion
The Hazelcast CLC simplifies the client setup process. Within minutes of installation, you can start to perform common tasks on clusters.
```

It looks like we are adding an install script for Windows.

Author: Updated as requested.

Author: Agreed to leave pending merging of PR for Windows version.

== Use Cases

> Example use cases for Hazelcast CLC.

Reviewer 2: ```suggestion
Example use cases for the Hazelcast CLC.
```

Author: Updated as requested.

=== Basic Cluster Management for Hazelcast Platform

Perform xref:hzc-cluster.adoc[administration tasks] on multiple clusters without writing code. Instead, use <<interactive-mode, interactive mode>> to explore available commands.

> Check or change the state of a cluster, update a configuration option, then switch to another cluster direct from the command line. For maximum flexibility, you can set cluster connection details at the command line or using a xref:configuration.adoc[custom configuration file]. 

Reviewer: ```suggestion
Check or change the state of a cluster, update a configuration option, then switch to another cluster directly from the command line. For maximum flexibility, you can set cluster connection details at the command line or using a xref:configuration.adoc[custom configuration file]. 
```

Author: Reworded slightly for easier reading.

NOTE: {hazelcast-cloud} has a separate link:{url-github-clc}[CLI for cluster management].

=== Access Data for Debugging

Get xref:hzc-map.adoc[direct access to map data] for quick debugging of prototypes or production applications. Easily read or query data on a cluster using <<sql-mode, SQL mode>> to view query results, and then perform updates without changing tools.

=== Scripting for Automation

Work in <<non-interactive-mode, non-interactive mode>> to script and execute repetitive administration, integration, or testing tasks.

== Modes

> Hazelcast CLC works in three modes:

Reviewer 2: ```suggestion
The Hazelcast CLC works in three modes:
```

Author: Updated as requested.

- Non-Interactive
- Interactive
- SQL

You can easily switch between modes to suit the context in which you're working.

> // Is this true, can you easily disable interactive mode? I couldn't. 

Reviewer 2: I assume you just use `exit` to get out of interactive mode. @utku-caglayan

Reviewer: Yes, from interactive mode one can exit by typing "exit" or pressing Ctrl+c together. Also you can switch to SQL Browser by typing "sql" from the interactive mode.

Reviewer 2: I think this is related to an issue #96

Author: I have added the ways to exit this mode and will track #96.

Reviewer: Oh I see, sorry to hear :(. I will try to reproduce the mentioned issue. Thanks

[[non-interactive-mode]]
=== Non-Interactive Mode

> This is the default mode for Hazelcast CLC. You need to manually enter and execute commands at the command line. This mode is useful for scripting.

Reviewer 2: ```suggestion
This is the default mode for the Hazelcast CLC. You need to manually enter and execute commands at the command line. This mode is useful for scripting.
```

Author: Updated as requested.

----
hzc map --cluster-name cluster1 --name myMap put --key myKey --value myValue
----

[[interactive-mode]]
=== Interactive Mode

Interactive mode uses auto-completion to display matching commands, subcommands, and parameters. This mode is useful for exploring CLC commands, and for manual tasks. 

image:ROOT:hzc-interactive-screenshot.png[HZC interactive]

In interactive mode, you can run the xref:hzc-map.adoc[`hzc map use` command] to avoid re-typing the object name on each command. In this example, the developer only needed to type `m1` once.

----
hzc
Connecting to the cluster ...
hzc localhost:5701@dev> map use m1
hzc localhost:5701@dev&m:m1> map put -k k1 -v v1
hzc localhost:5701@dev&m:m1> map get -k k1
v1
hzc localhost:5701@dev&m:m1>
----

You can xref:install-clc.adoc[enable interactive mode] for bash and zsh shells after installation. To start Hazelcast CLC in interactive mode, do the following:

[source,bash,subs="attributes+"]
----
hzc
----

[[sql-mode]]
=== SQL Mode
You can execute SQL statements in all modes, but SQL mode also gives you access to an interactive SQL browser to navigate query results.

image:ROOT:hzc-sql-browser-screenshot.png[HZC interactive]

> To start Hazelcast CLC in SQL mode, do the following:

Reviewer 2: ```suggestion
To start the Hazelcast CLC in SQL mode, do the following:
```

Author: Updated as requested.

[source,bash,subs="attributes+"]
----
hzc sql
----

== Limitations

> *Client authentication:* Hazelcast CLC supports only simple authentication for Hazelcast Platform. Hazelcast CLC does not currently support any other client authentication.

Reviewer 2: ```suggestion
*Client authentication:* Hazelcast CLC supports only simple authentication (username and password) for Hazelcast Platform. Hazelcast CLC does not currently support any other client authentication. If your Hazelcast Platform cluster uses other methods of client authentication, you cannot use Hazelcast CLC to connect to it.
```

Author: Updated as requested.

== Next Steps

> Install, configure, and explore Hazelcast CLC:

Reviewer 2: ```suggestion
Install, configure, and explore the Hazelcast CLC:
```

Author: Updated as requested.
